File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/analysis/cache/Actions.java
/*
 * Copyright 2015 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.drools.workbench.screens.guided.dtable.client.widget.analysis.cache;

import org.drools.workbench.screens.guided.dtable.client.widget.analysis.action.ActionInspector;
import org.drools.workbench.screens.guided.dtable.client.widget.analysis.action.ActionInspectorKey;
import org.drools.workbench.screens.guided.dtable.client.widget.analysis.checks.util.Conflict;
import org.drools.workbench.screens.guided.dtable.client.widget.analysis.checks.util.IsConflicting;
import org.drools.workbench.screens.guided.dtable.client.widget.analysis.checks.util.IsRedundant;
import org.drools.workbench.screens.guided.dtable.client.widget.analysis.checks.util.IsSubsuming;
import org.drools.workbench.screens.guided.dtable.client.widget.analysis.checks.util.Redundancy;

public class Actions
        extends MultiMap<ActionInspectorKey, ActionInspector>
        implements IsRedundant,
                   IsSubsuming,
                   IsConflicting {

    @Override
    public boolean isRedundant( Object other ) {
        if ( other instanceof IsSubsuming ) {
            return ( subsumes( other ) && ( (IsSubsuming) other ).subsumes( this ) );
        } else {
            return false;
        }
    }

    @Override
    public boolean subsumes( Object other ) {
        if ( other instanceof Actions ) {
            for ( ActionInspectorKey key : keys() ) {
                if ( !Redundancy.subsumes( get( key ),
                                           ( (Actions) other ).get( key ) ) ) {
                    return false;
                }
            }
            return true;
        } else {
            return false;
        }
    }

    @Override
    public boolean conflicts( Object other ) {
        if ( other instanceof Actions ) {
            for ( ActionInspectorKey key : keys() ) {
                if ( Conflict.isConflicting( get( key ),
                                             ( (Actions) other ).get( key ) ) ) {
                    return true;
                }
            }
            return false;
        } else {
            return false;
        }
    }

    public boolean hasValues() {
        for ( ActionInspector action : allValues() ) {
            if ( action.hasValue() ) {
                return true;
            }
        }
        return false;
    }
}

File: packages/pmml-editor/src/editor/reducers/Actions.ts
/*
 * Copyright 2020 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { StateControlActions, VersionActions } from "./PMMLReducer";
import { DataDictionaryActions } from "./DataDictionaryReducer";
import { DataDictionaryFieldActions } from "./DataDictionaryFieldReducer";
import { HeaderActions } from "./HeaderReducer";
import { ModelActions } from "./ModelReducer";
import { ScorecardActions } from "./ScorecardReducer";
import { CharacteristicsActions } from "./CharacteristicsReducer";
import { CharacteristicActions } from "./CharacteristicReducer";
import { AttributesActions } from "./AttributesReducer";
import { OutputActions } from "./OutputReducer";
import { OutputFieldActions } from "./OutputFieldReducer";
import { MiningSchemaActions } from "./MiningSchemaReducer";
import { MiningSchemaFieldActions } from "./MiningSchemaFieldReducer";

export type ActionMap<M extends { [index: string]: any }> = {
  [Key in keyof M]: M[Key] extends undefined
    ? {
        type: Key;
        payload: undefined;
      }
    : {
        type: Key;
        payload: M[Key];
      };
};

export enum Actions {
  Refresh = "REFRESH",
  Validate = "VALIDATE",
  SetVersion = "SET_VERSION",
  DeleteModel = "DELETE_MODEL",
  AddDataDictionaryField = "DATA_DICTIONARY_ADD_FIELD",
  AddBatchDataDictionaryFields = "DATA_DICTIONARY_BATCH_ADD",
  DeleteDataDictionaryField = "DATA_DICTIONARY_DELETE_FIELD",
  UpdateDataDictionaryField = "DATA_DICTIONARY_UPDATE_FIELD",
  ReorderDataDictionaryFields = "DATA_DICTIONARY_REORDER_FIELDS",
  SetHeaderDescription = "SET_HEADER_DESCRIPTION",
  AddOutput = "OUTPUT_ADD",
  AddBatchOutputs = "OUTPUT_BATCH_ADD",
  UpdateOutput = "OUTPUT_UPDATE",
  DeleteOutput = "OUTPUT_DELETE",
  AddMiningSchemaFields = "MINING_SCHEMA_ADD",
  DeleteMiningSchemaField = "MINING_SCHEMA_DELETE",
  UpdateMiningSchemaField = "MINING_SCHEMA_UPDATE",
  Undo = "UNDO",
  Redo = "REDO",
  Scorecard_SetModelName = "SCORECARD_SET_MODEL_NAME",
  Scorecard_SetCoreProperties = "SCORECARD_SET_CORE_PROPERTIES",
  Scorecard_AddCharacteristic = "SCORECARD_ADD_CHARACTERISTIC",
  Scorecard_DeleteCharacteristic = "SCORECARD_DELETE_CHARACTERISTIC",
  Scorecard_UpdateCharacteristic = "SCORECARD_UPDATE_CHARACTERISTIC",
  Scorecard_AddAttribute = "SCORECARD_ADD_ATTRIBUTE",
  Scorecard_DeleteAttribute = "SCORECARD_DELETE_ATTRIBUTE",
  Scorecard_UpdateAttribute = "SCORECARD_UPDATE_ATTRIBUTE"
}

export type AllActions =
  | StateControlActions
  | VersionActions
  | HeaderActions
  | MiningSchemaActions
  | MiningSchemaFieldActions
  | DataDictionaryActions
  | DataDictionaryFieldActions
  | ModelActions
  | ScorecardActions
  | CharacteristicsActions
  | CharacteristicActions
  | AttributesActions
  | OutputActions
  | OutputFieldActions;
